### PR TITLE
Droplets for tiled detectors.

### PR DESCRIPTION
**Pull request by silkenelson**
Enable droplets to work on tiled detectors. Also some work on the imageFunc so that you could make an image from droplets & the sparsify instead of having to do it the other way around.

### CODE REVIEW

File: smalldata_tools/ana_funcs/azimuthalBinning.py [hunk 1]
```diff
@@ -97,13 +97,18 @@ def setFromFunc(self, func=None):
...
+        maskattr='_mask'
+        if not hasattr(func, maskattr):
+            maskattr='mask'
+            if not hasattr(func, maskattr):
+                maskattr=None
+        if maskattr is not None and getattr(func,maskattr) is not None:
+            if isinstance(func, ROIFunc): 
+                self._mask = getattr(func, maskattr).astype(bool).flatten()
+            else:
+                self._mask = (~(getattr(func,maskattr).astype(bool))).flatten()
```
Comment (vespos) [inline smalldata_tools/ana_funcs/azimuthalBinning.py:111]: Are masks really defined with a different convention for ROIFunc compared to all other functions? IF yes we should probably make that uniform...

File: smalldata_tools/ana_funcs/roi_rebin.py [hunk 2]
```diff
@@ -398,22 +401,39 @@ def setFromDet(self, det):
...
+                        self.__dict__['i%s'%coord] = getattr(det, 'i%s'%coord).flatten()
+                    except:
+                        #
+                        # DEBUG ME: looks right, but not sure
+                        #
+                        intcoord = self.__dict__[coord].copy()
+                        intcoord = intcoord - np.nanmin(intcoord)
+                        intcoord = (intcoord/np.nanmax(intcoord)*self.imgShape[0]).astype(int)
+                        self.__dict__['i%s'%coord] = intcoord
+            else:
```
Comment (vespos) [inline smalldata_tools/ana_funcs/roi_rebin.py:421]:
```suggestion
            elif isinstance(self._coords, dict)
```

**Review by vespos — COMMENTED** [review]
Some comments. Apart from the measurements.label thing (perhaps I am missing smth there), the rest is not critical

File: smalldata_tools/ana_funcs/roi_rebin.py [hunk 3]
```diff
@@ -366,14 +368,15 @@ class imageFunc(DetObjectFunc):
...
-        self.coords = kwargs.get('coords',None)
+        self._coords = kwargs.get('coords',None)
         self.imgShape = kwargs.get('imgShape',None)
         self.correction = kwargs.get('correction',None)
         self.mask = kwargs.get('mask',None)
         if self.mask is not None: self.mask = np.asarray(self.mask,dtype=np.bool).flatten()
 
     def setFromDet(self, det):
         super(imageFunc, self).setFromDet(det)
+        self._pedShape = det.ped.shape
```
[the remark this answers is not in the document]
Comment (silkenelson) [reply smalldata_tools/ana_funcs/roi_rebin.py:379]: While I sort of understand the concern, just adding a variable called "pedShape" does not change anything here really as it's the shape of the pedestal....

File: smalldata_tools/ana_funcs/azimuthalBinning.py (hunk 1 above)
Comment (silkenelson) [reply smalldata_tools/ana_funcs/azimuthalBinning.py:111]: I basically agree with this: I had found the inversion already there and essentially extended it. In the end, I believe this boils down to the fact that masks as used for masked arrays usually are '1' for to-be-masked entries while many other pieces of code returns masks to be used multiplicatively (including I believe the psana.Detector functions). This was fairly hard to get correct through all the chains, so proceed with care and many tests if you want to take this one.

File: smalldata_tools/ana_funcs/droplet.py [hunk 4]
```diff
@@ -185,15 +186,18 @@ def dropletize(self, data):
...
-                dat_dict['col']=pos_drop[:,1]
-            dat_dict['npix']=npix_drop
+                dat_dict['row']=pos_drop[:,pos_drop.shape[1]-2]
+                dat_dict['col']=pos_drop[:,pos_drop.shape[1]-1]
+                dat_dict['tile']=pos_drop[:,0]
         else:
+            #this should be tested for tiled detectors!
             #t2 = time.time()
             self.regions = measure.regionprops(imgDrop, intensity_image=img, cache=True)
             dropSlices = measurements.find_objects(imgDrop)
```
[the remark this answers is not in the document]
Comment (silkenelson) [reply smalldata_tools/ana_funcs/droplet.py:203]: Ok that worries me too then: if you have confirmed that the default behavior works wrong, then we need to change the  passed-in shape when we have a 3-d detector. Maybe we should also make a note to check what the span photon code does: I only know it returns the correct shape, but do not know if it knows how to work with a 3-d input correctly. Do you want to take it or should I take a look and push a change to the same branch?

File: smalldata_tools/ana_funcs/roi_rebin.py [hunk 5]
```diff
@@ -398,22 +401,39 @@ def setFromDet(self, det):
...
-                    # DEBUG ME: looks right, but not sure
-                    #
-                    intcoord = self.__dict__[coord].copy()
-                    intcoord = intcoord - np.nanmin(intcoord)
-                    intcoord = (intcoord/np.nanmax(intcoord)*self.imgShape[0]).astype(int)
-                    self.__dict__['i%s'%coord] = intcoord
-
-            self._coordTuple = tuple( self.__dict__['i%s'%coord] for coord in self.coords)
-            self._n_coordTuple = tuple( np.max(self.__dict__['i%s'%coord]+1) for coord in self.coords)
+        if isinstance(self._coords, list) or isinstance(self._coords, dict):
```
[the remark this answers is not in the document]
Comment (silkenelson) [reply smalldata_tools/ana_funcs/roi_rebin.py:404]: Ok, I think this is a remnant from when I thought I could get away with less indentation or less lines of code.

[the remark this answers is not in the document]
Comment (silkenelson) [reply smalldata_tools/ana_funcs/roi_rebin.py:404]: I actually don't think it works how you suggest to do it unless I just don't see it.. From line 435, code is shared for both list & dict cases for coords as input. I also can't request that coords exist as there are other use cases below. If I'm using else for the 'if list' after the 'if list or dict' of elif 'dict' does not matter and I find the else better.

**Comment by vespos** [comment]
Looks good now, feel free to merge.
Not sure I'll have time to go over the photon function this week, but I can try next one. We can keep it into that branch I guess

File: smalldata_tools/ana_funcs/droplet.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (silkenelson) [reply smalldata_tools/ana_funcs/droplet.py:207]: Update: I actually fixed this and must have forgotten to push these changes to the branch. Need to set a reminder.

**Pull request merged**